ui/app/helpers/parse-pki-cert.js

```diff
@@ -17,44 +17,31 @@ export function parseCertificate(certificateContent) {
       can_parse: false,
     };
   }
-  // We wish to get the CN element out of this certificate's subject. A
-  // subject is a list of RDNs, where each RDN is a (type, value) tuple
-  // and where a type is an OID. The OID for CN can be found here:
-  //
-  //    http://oid-info.com/get/2.5.4.3
-  //    https://datatracker.ietf.org/doc/html/rfc5280#page-112
-  //
-  // Each value is then encoded as another ASN.1 object; in the case of a
-  // CommonName field, this is usually a PrintableString, BMPString, or a
-  // UTF8String. Regardless of encoding, it should be present in the
-  // valueBlock's value field if it is renderable.
-  const commonNameOID = '2.5.4.3';
-  const commonNames = cert?.subject?.typesAndValues
-    .filter((rdn) => rdn?.type === commonNameOID)
-    .map((rdn) => rdn?.value?.valueBlock?.value);
-
-  // Theoretically, there might be multiple (or no) CommonNames -- but Vault
-  // presently refuses to issue certificates without CommonNames in most
-  // cases. For now, return the first CommonName we find. Alternatively, we
-  // might update our callers to handle multiple, or join them using some
-  // separator like ','.
-  const commonName = commonNames ? (commonNames.length ? commonNames[0] : null) : null;
-
-  // Date instances are stored in the value field as the notAfter/notBefore
-  // field themselves are Time values.
+
+  const subjectParams = parseSubject(cert?.subject?.typesAndValues);
   const expiryDate = cert?.notAfter?.value;
   const issueDate = cert?.notBefore?.value;
-  const serialNumber = Convert.ToHex(cert.serialNumber.valueBlock.valueHex)
-    .match(/.{1,2}/g)
-    .join(':');
+
+  // TODO wrap to catch errors, only parse if cross-signing?
+  // for cross-signing
+  const { alt_names, uri_sans } = parseExtensions(cert?.extensions);
+  const [signature_bits, use_pss] = mapSignatureBits(cert?.signatureAlgorithm);
```

Comment:
@cipherboy can we use `signature_bits` to get the `key_bits` and `key_type` info?

Reply:
I have good news, and bad news. :-)

The bad news is, sorry that won't work. The signature's key type (RSA/...) is based on the parent issuer's CA type, which we're changing here (from an old parent issuer to a new one to form the cross-signed chain).

The good news is, since we're using an existing key (`intermediate/generate/existing`), you don't actually need to pass `key_type` and `key_bits` like @kitography does (it already exists in Vault). Kit's work is also implementing reissuance and not just cross-signing (the former can rotate key material, the latter leaves it alone).

The `use_pss` and `signature_bits` are still necessary though to tell the new parent CA what to do. :-)

Reply:
Ah that makes sense! I totally should have realized that when I was playing around with `key_type` and `key_bit` I had to use a different endpoint, which is _not_ what we use for cross-signing - thank you for the reminder!! (I should really not work after 8:30pm 🥲 )

YAY, I love the sound of not having to pass them 🎉 😄